velox/functions/sparksql/aggregates/FirstLastAggregate.cpp

```diff
@@ -0,0 +1,427 @@
+    if (numeric) {
+      BaseAggregate::doExtractValues(
+          groups, numGroups, result, [&](char* group) {
+            auto accumulator =
+                exec::Aggregate::value<TAccumulator<TDataType>>(group);
+            return accumulator->value_;
+          });
+    } else {
+      VELOX_CHECK(result);
+      (*result)->resize(numGroups);
+
+      auto* rawNulls = exec::Aggregate::getRawNulls(result->get());
+
+      for (auto i = 0; i < numGroups; ++i) {
+        char* group = groups[i];
+        if (exec::Aggregate::isNull(group)) {
+          (*result)->setNull(i, true);
+        } else {
+          exec::Aggregate::clearNull(rawNulls, i);
+          auto accumulator =
+              exec::Aggregate::value<TAccumulator<SingleValueAccumulator>>(
+                  group);
+          accumulator->value_.read(*result, i);
+        }
+      }
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    extractValues(groups, numGroups, result);
+  }
+
+  void destroy(folly::Range<char**> groups) override {
+    if (!numeric) {
+      for (auto group : groups) {
+        auto accumulator =
+            exec::Aggregate::value<TAccumulator<SingleValueAccumulator>>(group);
+        accumulator->value_.destroy(exec::Aggregate::allocator_);
+      }
+    }
+  }
+};
+
+template <bool ignoreNull, typename TDataType, bool numeric>
+class FirstAggregate
+    : public FirstLastAggregateBase<numeric, SimpleAccumulator, TDataType> {
+ public:
+  explicit FirstAggregate(TypePtr resultType)
+      : FirstLastAggregateBase<numeric, SimpleAccumulator, TDataType>(
+            resultType) {}
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    DecodedVector decoded(*args[0], rows);
+
+    rows.applyToSelected(
+        [&](vector_size_t i) { updateValue(i, groups[i], decoded); });
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    addRawInput(groups, rows, args, mayPushdown);
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    DecodedVector decoded(*args[0], rows);
+
+    rows.testSelected(
+        [&](vector_size_t i) { return updateValue(i, group, decoded); });
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    addSingleGroupRawInput(group, rows, args, mayPushdown);
+  }
+
+ private:
+  // If we found a valid value, set accumulator valid flags to true, then skip
+  // remaining rows in group.
+  bool updateValue(vector_size_t i, char* group, DecodedVector& decoded) {
```

Comment:
Maybe this can be simplified a bit. See https://github.com/facebookincubator/velox/compare/main...mbasmanova:velox-1:first_last_agg?expand=1